software/py/blood_graph.py

```diff
@@ -37,282 +39,322 @@ class BloodGraph:
-                                             "stall_amo_aq"                           : (0x00, 0x00, 0x00), ## black
-                                             "stall_amo_rl"                           : (0x00, 0x00, 0x00), ## black
 
-                                             "icache_miss"                            : (0x00, 0x00, 0xff), ## blue
-                                             "stall_ifetch_wait"                      : (0x00, 0x00, 0xff), ## blue
-                                             "bubble_icache"                          : (0x00, 0x00, 0xff), ## blue
+    # parses vanilla_operation_trace.csv to generate operation traces
+    def __parse_traces(self, trace_file):
+        traces = []
+        with open(trace_file) as f:
+            csv_reader = csv.DictReader(f, delimiter=",")
+            for row in csv_reader:
+                trace = {}
+                trace["x"] = int(row["x"])  
+                trace["y"] = int(row["y"])  
+                trace["operation"] = row["operation"]
+                trace["cycle"] = int(row["cycle"])
+                traces.append(trace)
+        return traces
+
+
+    # Parses vanilla_stats.csv to generate timing stats 
+    # to gather start and end cycle of entire graph
+    def __parse_stats(self, stats_file):
+        stats = []
+        if(stats_file):
+            if (os.path.isfile(stats_file)):
+                with open(stats_file) as f:
+                    csv_reader = csv.DictReader(f, delimiter=",")
+                    for row in csv_reader:
+                        stat = {}
+                        stat["global_ctr"] = int(row["global_ctr"])
+                        stat["time"] = int(row["time"])
+                        stats.append(stat)
+            else:
+                warnings.warn("Stats file not found, overriding blood graph's start/end cycle with traces.")
+        return stats
 
-                                             "bubble_branch_mispredict"               : (0x00, 0x00, 0x00), ## black
-                                             "bubble_jalr_mispredict"                 : (0x00, 0x00, 0x00), ## black
-                                             "bubble_fp_op"                           : (0x00, 0x00, 0x00), ## black
-                                             "bubble"                                 : (0x00, 0x00, 0x00), ## black
 
-                                             "stall_md"                               : (0xff, 0xff, 0xff), ## white
-                                           }
-        self.abstract_unified_instr_color    =                                          (0xff, 0xff, 0xff)  ## white
-        self.abstract_unified_fp_instr_color =                                          (0xff, 0xff, 0xff)  ## white
+    # look through the input file to get the tile group dimension (x,y)
+    def __get_tile_group_dim(self, traces):
+        xs = list(map(lambda t: t["x"], traces))
+        ys = list(map(lambda t: t["y"], traces))
```

Comment:
Educational moment: 

Instead of iterating over the list twice, you can do a list comprehension: 

`[t["x"] for t in traces]` 

accomplishes the same thing and is usually faster

Reply:
Oh yeah that's a good point! I fixed it, thanks!